### PR TITLE
[WIP] Place unroll -> iterate refactor

### PR DESCRIPTION
**Pull request by spastorino**
**Do not merge this** is work in progress.

r? @oli-obk

### CODE REVIEW

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -2044,6 +2044,86 @@ impl<'tcx> Place<'tcx> {
...
+    /// ```
+    fn iter(&self) -> PlaceProjectionsIter<'_, 'tcx> {
+        PlaceProjectionsIter { value: self }
+    }
+}
+
+/// Iterator over components; see `PlaceProjections::iter` for more
+/// information.
+///
+/// N.B., this is not a *true* Rust iterator -- the code above just
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: What we could do is make it a real iterator and also implement https://doc.rust-lang.org/std/iter/trait.FusedIterator.html for that iterator. This way we guarantee that after the first `None` we keep getting `None`s.

**Comment by oli-obk** [comment]
@spastorino this looks like a good intermediate step that we could merge before any further changes.

**Comment by Dylan-DPC** [comment]
ping from triage @oli-obk any updates?

**Comment by spastorino** [comment]
@Dylan-DPC I'm the one "working" on this. This is still in WIP, we need to land some stuff before. I'm slowly catching up with stuff but have been very busy.

**Comment by Dylan-DPC** [comment]
thanks. I thought some part of the review was pending. Great :)

**Comment by Centril** [comment]
Ping from triage, @spastorino What other stuff is this blocked on? what PRs if any?

**Comment by spastorino** [comment]
@Centril blocked on https://rustlatam.org :joy:. Now that the thing have finished I'm back to work on `Place 2.0`

**Comment by spastorino** [comment]
Going to close this for now and reopen after I change `Place` to a struct

**Pull request closed without merging**